src/ConfigGen/Program.cs

```diff
@@ -418,26 +418,40 @@ static string createClassHeader(string name, string docs, bool derive)
             return codeText;
         }
 
-        static string createConsumerSpecific()
+        static string createClassConstructors(string name)
         {
-            return
-@"        /// <summary>
-        ///     Initialize a new empty <see cref=""ConsumerConfig"" /> instance.
+            var codeText = "\n";
+            codeText += $@"        /// <summary>
+        ///     Initialize a new empty <see cref=""{name}"" /> instance.
         /// </summary>
-        public ConsumerConfig() {}
+        public {name}() {{ }}
 
         /// <summary>
-        ///     Initialize a new <see cref=""ConsumerConfig"" /> instance based on
+        ///     Initialize a new <see cref=""{name}"" /> instance based on
         ///     an existing <see cref=""ClientConfig"" /> instance.
         /// </summary>
-        public ConsumerConfig(ClientConfig config) { this.properties = new Dictionary<string, string>(config.ToDictionary(a => a.Key, a => a.Value)); }
+        public {name}(ClientConfig config) : base(config) {{ }}
 
         /// <summary>
-        ///     Initialize a new <see cref=""ConsumerConfig"" /> instance based on
+        ///     Initialize a new <see cref=""{name}"" /> instance based on
         ///     an existing key/value pair collection.
         /// </summary>
-        public ConsumerConfig(IEnumerable<KeyValuePair<string, string>> config) { this.properties = new Dictionary<string, string>(config.ToDictionary(a => a.Key, a => a.Value)); }
+        public {name}(IEnumerable<KeyValuePair<string, string>> config) : base(config) {{ }}
+
+        /// <summary>
+        ///     Initialize a new <see cref=""{name}"" /> wrapping
+        ///     an existing key/value pair collection.
+        /// </summary>
+        public {name}(IDictionary<string, string> config) : base(config) {{ }}

```

Comment:
The reasoning behind using `IEnumerable<KeyValuePair<string, string>>` is that it's the most general, so covers the most cases. In particular, `IDictionary<string,string>` extends this. Why is another constructor overload necessary?

That said, people would find `IDictionary` more friendly, and i might have made an error in judgement if `IEnumerable` isn't useful in practice (i can't recall).

Reply:
Oh, if you pass in an `IDictionary` , then there's no copy so you could modify the dictionary independently and the changes would be reflected in the config classes. This does seem like it could be useful and a good change. But I don't like that it works like that is you pass a dictionary and doesn't if you pass an `IEnumerable`. I think it would be better to change the get/set methods to work with the IEnumerable rather than a dictionary and change the `properties` field to be an `IEnumerable`. This would be O(N) of course, but efficiency doesn't matter here at all, and linear scan isn't going to be meaningfully slow anyway since there's so few items. As an added bonus, there are less constructor overloads.

Reply:
indeed this is the idea behind this ctor ie values are changed in-place. I'll add some xmldoc to highlight this difference.

It is not possible to do the same thing with the IEnumerable.
KeyValuePair is a "value type", is it's immutable. It is not possible to change the values in-place.

So the ctors that take IEnumerable _must_ do a copy

Reply:
I've also made the `Config(Config ...)` ctor behave "in-place" too. It is a change in approach but I don't think this will affect any "normal" code code/samples I've seen.

Also added a `Clone()` helper method for when someone _really_ needs the "copy" style.

Reply:
ahh crap, yeah, ok - your modifications make sense.

Reply:
it's the weekend and i have't checked this super carefully, but a few comments and then it's looking pretty good to me. thanks for your efforts!

Reply:
I think the other changes have resolved this?